doc/guidelines.md

```diff
@@ -0,0 +1,242 @@
+
+If these guidelines are adopted, then the expectation is that all components would follow them.
+This reduces cognitive load when using components -- no need to remember one component's peculiarities.
+It also allows Agent-wide changes, where we make the same formulaic change to each component.
+If a situation arises that contradicts the guidelines, then we can update the guidelines (and change all affected components).
+
+## Component Definition
+
+A component is defined in a dedicated package under `comp/`, with the following defined in `component.go`:
+
+ * Extensive package-level documentation.
+   This should define, as precisely as possible, the behavior of the component, acting as a contract on which users of the component may depend.
+
+   The documentation (both package-level and method-level) should also include:
+
+   * Precise information about which interface methods can be called during the setup phase, and which must only be called after the component is started.
+   * Precise information about data ownership of passed values and returned values.
+     By default, any mutable value returned by a component will not be modified after it is returned.
+     Similarly, any mutable value passed to a component will not be later modified either by the component or the caller.
+     Any deviation from these defaults should be clearly documented.
+   * Precise information about goroutines and blocking (does the method block? is a callback invoked in a dedicated goroutine? what happens if a channel is full?).
+     By default, all methods are assumed to return without blocking.
+
+ * `pkg.Component` -- the type implemented by the component.
+   This can be an empty interface, but is the type by which other components will find this one via `fx` and so must still appear in a function signature.
+   It should have a formulaic doc string like `// Component is the component type.`, deferring documentation to the package docs.
+   All interface methods should be exported and thoroughly documented.
+
+ * `pkg.Module` -- an `fx.Option` that can be included in an `fx.App` to make this component available.
+   To improve logging, use `fx.Module(comppath, ..)`, where `comppath` is the root-relative package path, e.g., `"comp/util/log"`.
+   It should have a formulaic doc string like `// Module defines the fx options for this component.`
+
+Components should not be nested; that is, no component's Go path should be a prefix of another component's Go path.
+
+### Implementation
+
+The completed `component.go` looks like this:
+
+```go
+// Package foo ... (detailed doc comment for the component)
+package config
+
+// team: some-team-name
+
+const componentname = "comp/foo" // ... (should match Go package name)
+
+// Component is the component type.
+type Component interface {
+	// Foo is ... (detailed doc comment)
+	Foo(key string) string
+}
+
+// Module defines the fx options for this component.
+var Module = fx.Module(
```

Comment:
Completing https://github.com/djmitche/dd-agent-comp-experiments/pull/1#discussion_r934263930 

`fx.Module` will include some dependencies but not all. For example, `fx.Module` would not include external dependencies such as `config` or `telemetry`. It means when a user try to instantiate a single component, they need to know what are these external dependencies.
 
I can think of several way to handle this question:
1. Don't try to list these external dependencies. They will be caught anyway by https://github.com/djmitche/dd-agent-comp-experiments/pull/1#discussion_r936366865 (Don't know if the error message will be usable or not).   I am wondering if this solution scales well (and what would be the limit and will the Agent repo will reach this limit).
2. Make explicit the external dependencies solution 1. The main provider (`newFoo` in this example) can have 2 parameters:  a `dependencies` struct that list external dependencies and `internalDependencies` struct that list dependencies provides by `fx.Module`
3. Make explicit the external dependencies solution 2. Use a factory method. `populateOptions` allow to override the default provider.
```go
var Module = fx.Module(
componentName,
    fx.Provide(newInternalDeps), // Register only internal deps
)

var Factory = func(externalDependencies dependencies, overrideOptions ...fx.Option) Component {
        var options []fx.Option
        var component Component 

	options = append(options, fx.Provide(....))
        options = append(options, overrideOptions)       
	options = append(options, fx.Populate(&component))
	app := fx.New(options...)		
	if err := app.Start(context.Background()); err != nil {
		panic(err)
	}
	defer app.Stop(context.Background())
        return component
})
```

For `2.` and `3.` a unit test should check that external dependencies are correct using `ValidateApp` . See https://github.com/djmitche/dd-agent-comp-experiments/pull/1#discussion_r936366865.

`3.` requires more code (an helper function can minimize that) but provides a better locality as internal dependencies cannot interfere with other components (For very big project this is very important). 

What are your thoughts?

Reply:
I think nesting apps within one another is dangerous territory -- I don't think those will be able to share components with the "parent" app very easily.

I think we could get away with something like (2) if Fx was OK with having the same option appear multiple times.  That is, if each component could list the modules it depends on like `config.Module, log.Module, hostname.Module, ..`  and even validate that.  Sadly, Fx fails for this, and it doesn't make it easy to introspect fx.Options so that we might de-duplicate them manually.

We've mentioned a "core" bundle before, and it may be that we can just require that all bundle-level dependencies be expressed at the bundle level (so, comp/dogstatsd bundle depends on the comp/core and comp/forwarder bundles).  That might mean including some modules that aren't ultimately required, but that only matters to binary size (no runtime impact).  It'd also be easy to validate.

I am going to try out the last idea first, and see where we get to.

Reply:
> I think nesting apps within one another is dangerous territory -- I don't think those will be able to share components with the "parent" app very easily.

Consider an `Aggregator` component: the factory function registers only the inner dependencies of Aggregator (The ones that were in `/pkg/aggregator`). All other components are injected as a parameter. I think it should be straightforward in practice even if it makes a small room for issue. The main pros is isolation especially for very big code base. The code of the agent is probably short enough so we can probably skip this idea for now.